File: _8_Remove_Element.java
import java.util.Arrays; // Para imprimir arrays fácilmente

public class _8_Remove_Element {
    public static void main(String[] args) {
        Solution sol = new Solution();
        int[] nums1 = {3,2,2,3};
        int k1 = sol.removeElement(nums1, 3);
        System.out.println("k1 = " + k1 + ", nums1 = " + Arrays.toString(Arrays.copyOf(nums1, k1)));
        // Esperado: [0, 1]
        int[] nums2 = {0,1,2,2,3,0,4,2};
        int k2 = sol.removeElement(nums2, 2);
        System.out.println("k2 = " + k2 + ", nums2 = " + Arrays.toString(Arrays.copyOf(nums1, k2)));

        // Esperado: [1, 2]
        int[] nums3 = {3, 3};
        int k3 = sol.removeElement(nums3, 2);
        System.out.println("k3 = " + k3 + ", nums3 = " + Arrays.toString(Arrays.copyOf(nums1, k3)));


    }
}
// My solution
// class Solution {
//     public int removeElement(int[] nums, int val) {
//         int count = 0;
//         int tam = nums.length;
//         for (int num : nums) {
//         if (num == val) count++;
//         }
//         if (count == 0) return tam;

//         for (int i = 0; i < tam; i++) {
//             int c = i+1;
//             while (nums[i] == val) {
//                 if (c >= tam){
//                     break;
//                 }
//                 int aux = nums[i];
//                 nums[i] = nums[c]; 
//                 nums[c] = aux;
//                 c++;
//             }
//         }
        
//         return tam-count;
//     }
// }


// Other solution
class Solution {
    public int removeElement(int[] nums, int val) {
        int j = 0; 
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] != val) {
                nums[j] = nums[i];
                j++;
            }
        }
        return j; 
    }
}

File: _8_Remove_Element.ts

// My solution
// function removeElement(nums: number[], val: number): number {
//     let j = 0;
//     for (let i = 0; i < nums.length; i++) {
//         if (nums[i] !== val) {
//             nums[j] = nums[i];
//             j++;
//         }
//     }
//     return j; 
// }

//Other solution
function removeElement(nums: number[], val: number): number {
    let l = 0, r = nums.length - 1;
    
    while (l <= r) {
        if (nums[l] !== val) {
            l++;
        } else {
            while (nums[r] === val) {
                r--;
            }

            if (l > r) break;

            nums[l] = nums[r];
            l++;
            r--;
        }
    }

    return l;
};

let nums4 = [3, 2, 2, 3];
let k4 = removeElement(nums4, 3);
console.log("k1 =", k4, ", nums1 =", nums4.slice(0, k4));

let nums5 = [0, 1, 2, 2, 3, 0, 4, 2];
let k5 = removeElement(nums5, 2);
console.log("k2 =", k5, ", nums2 =", nums5.slice(0, k5));

let nums6 = [3, 3];
let k6 = removeElement(nums6, 3);
console.log("k3 =", k6, ", nums3 =", nums6.slice(0, k6));

